翻译Marionette动画系统 (#1905)
cocos-creator/creator-docs: en/animation/marionette/state-translation.md

# State Transitions

> **Transition** represents a transition between two states, which we can classify depending on the source where the transition occurs.

Reviewer: ```suggestion
**Transition** represents a transition between two states, which can be classified depending on the source where the transition occurs.
```

![condition-properties](state-translation/condition-properties.png)

- **Duration**: set the period to complete the current transition, so the animation will switch smoothly. The unit is **seconds** or **times**, default is 0.3 seconds, you can click the unit after the input box to switch.

    ![change duration](state-translation/change-duration.gif)

    - When set to **seconds** (default): indicates that the state transition is completed within the set number of seconds

    - When set to **times**: means the period of the state transition is based on the number of times the transition source state is played. For example, if the animation length of the transition source state is 3 s, when **Duration** is set to 0.9 times, it means the transition period is 0.9 ✖️ 3s = 2.7s.

- **Enable Exit Times**: indicates whether the transition source state ends playing as one of the transition conditions. If checked, the transition needs to occur to meet the **Exit Times** condition set below. Play count is used as a special condition, when play count is enabled, the transition will happen only when both the play count condition and the transition condition are satisfied.

> - **Exit Times**: sets the total amount of time the animation plays before the transition source state begins (Total = Exit Times ✖️ Animation Length), defaults to 1 time, and takes effect only when **Enable Exit Times** is checked. Note that animations that do not loop will stop when the first play is completed and wait for the remaining number of plays to complete before starting the transition. If you need to loop the animation automatically according to the number of times, please set the wrap mode in the skeleton animation asset beforehand. For example:

Reviewer: ```suggestion
- **Exit Times**: sets the total amount of time the animation plays before the transition source state begins (Total = Exit Times ✖️ Animation Length), defaults to 1 time, and takes effect only when **Enable Exit Times** is checked. It is important to note that animations that do not loop will stop when the first play is completed and wait for the remaining number of plays to complete before starting the transition. To loop the animation automatically according to the number of times, please set the wrap mode in the skeleton animation asset beforehand. For example:
```

    - If the animation of the transition source is an **non-loop animation**, the playing time is 3 s, **Exit Times** is set to 1.4, then the animation of the transition source will continue to stay in the last frame for 1.2s (3s × 0.4) after playing for 3s, before starting the transition.

      ![non-loop](state-translation/non-loop.png)

    - If the animation of the transition source is an **loop animation**, playing time is 3 s, **Exit Times** is set to 1.4, then the animation of the transition source will continue to loop the animation to 1.2s (3s × 0.4) after playing 3s, before starting the transition.

      ![loop](state-translation/loop.png)

- **Condition**: there are certain conditions that need to be met when setting up transitions that occur between states, as described below:

> Without using other transition conditions, if you want the animation of the source state to gradually switch to other states when it is almost finished playing once, and the animation of the source state is just finished playing when the switch is done, then we can set **Duration** and **Exit Times** to achieve this, it is recommended to set the unit of **Duration** to **times**, and then make the sum of **Duration** and **Exit Times** sum to **1**.

Reviewer: ```suggestion
Without using other transition conditions, for the animation of the source state to gradually switch to other states when it is almost finished playing once, and the animation of the source state is just finished playing when the switch is done, then we can set **Duration** and **Exit Times** to achieve this, it is recommended to set the unit of **Duration** to **times**, and then make the sum of **Duration** and **Exit Times** sum to **1**.
```

- **Trigger condition**: condition is satisfied when the trigger variable is triggered.

  ![trigger-condition](state-translation/trigger-condition.png "trigger condition")

**Transition supports specifying multiple transition conditions at the same time, and the transition will occur when and only when all conditions are satisfied.**

### Unconditional Transitions

For **ordinary transitions**, it is allowed to not specify any transition conditions, for example, many times **Entry** needs to transition to other states unconditionally.

However, for **animation transitions**, transitions where neither **conditions** nor **Exit Times** conditions are specified are meaningless and the Creator will simply ignore the transition.

## An edge case

There are cases that cause a state machine to stay on **Entry** or **Exit**, e.g.

- Transitioned to the **Entry** of a sub-state machine, but there is no transition in the sub-state machine that would satisfy the condition.

- Transitioned to the **Exit** of a sub-state machine, but there is no transition in the parent state machine that would satisfy the condition for that sub-state machine.

> We call this situation **state machine hover**.

Reviewer: ```suggestion
This situation is called **state machine hover**.
```
